main.tf

```diff
@@ -0,0 +1,148 @@
+/**
+ * Creates a lambda function with associated role and policies, which
+ * will log to Cloudwatch Logs.
+ *
+ * Creates the following resources:
+ *
+ * * Lambda function
+ * * IAM role and policies described in the inputs
+ * * Cloudwatch Logs group
+ *
+ * ## Usage
+ *
+ * ```hcl
+ * module "my_lambda_function" {
+ *   source                = "trussworks/terraform-aws-lambda"
+ *   name                  = "my_app"
+ *   job_identifier        = "instance_alpha"
+ *   runtime               = "go1.x"
+ *   lambda_role_policy    = "${data.aws_iam_policy_document.my_app_policy.json}"
+ *   s3_bucket             = "my_s3_bucket"
+ *   s3_key                = "my_app/1.0/my_app.zip"
+ *
+ *   subnet_ids            = ["subnet-0123456789abcdef0"]
+ *   security_group_ids    = ["sg-0123456789abcdef0"]
+ *
+ *   source_types          = ["events"]
+ *   source_arns           = ["${aws_cloudwatch_event_rule.trigger.arn}"]
+ *
+ *   env_vars {
+ *     VARNAME = "value"
+ *   }
+ *
+ *   tags {
+ *     "Name" = "my_app_lambda"
+ *   }
+ *
+ * }
+ * ```
+ */
+
+data "aws_region" "current" {}
+data "aws_caller_identity" "current" {}
+
+# This is the IAM policy for letting lambda assume roles.
+data "aws_iam_policy_document" "assume_role" {
+  statement {
+    effect = "Allow"
+
+    actions = ["sts:AssumeRole"]
+
+    principals {
+      type        = "Service"
+      identifiers = ["lambda.amazonaws.com"]
+    }
+  }
+}
+
+# Create the IAM role for the Lambda instance.
+resource "aws_iam_role" "main" {
+  name               = "lambda-${var.name}-${var.job_identifier}"
+  assume_role_policy = "${data.aws_iam_policy_document.assume_role.json}"
+}
+
+# Create and attach the user-provided policy to the above IAM role.
+resource "aws_iam_role_policy" "main" {
+  name = "lambda-${var.name}-${var.job_identifier}"
+  role = "${aws_iam_role.main.id}"
+
+  policy = "${var.lambda_role_policy}"
```

Comment:
It feels like the user-provided policy should be optional, in which case this inline policy should be the CloudWatch Logs policy that every lambda needs.

Reply:
I went back and forth on these a bunch...I think what I ended up picking was kind of arbitrary. No matter what I decided I felt like I was locking something in...but I think what you're saying makes perfect sense.